src/sfloader/fluid_samplecache.c

```diff
@@ -0,0 +1,279 @@
+static fluid_samplecache_entry_t *get_samplecache_entry(SFData *sf, unsigned int sample_start, unsigned int sample_count);
+static void delete_samplecache_entry(fluid_samplecache_entry_t *entry);
+
+static int fluid_get_file_modification_time(char *filename, time_t *modification_time);
+
+
+/* PUBLIC INTERFACE */
+
+int fluid_samplecache_load(SFData *sf,
+                           unsigned int sample_start, unsigned int sample_count,
+                           int try_mlock, short **sample_data, char **sample_data24)
+{
+    fluid_samplecache_entry_t *entry;
+    int ret;
+
+    fluid_mutex_lock(samplecache_mutex);
+
+    entry = get_samplecache_entry(sf, sample_start, sample_count);
+    if (entry == NULL)
+    {
+        entry = new_samplecache_entry(sf, sample_start, sample_count);
+        if (entry == NULL)
+        {
+            ret = FLUID_FAILED;
+            goto unlock_exit;
+        }
+
+        samplecache_list = fluid_list_prepend(samplecache_list, entry);
+    }
+
+    if (try_mlock && !entry->mlocked)
+    {
+        /* Lock the memory to disable paging. It's okay if this fails. It
+         * probably means that the user doesn't have the required permission. */
+        if (fluid_mlock(entry->sample_data, entry->sample_count * 2) == 0)
+        {
+            if (entry->sample_data24 != NULL)
+            {
+                entry->mlocked = (fluid_mlock(entry->sample_data24, entry->sample_count) == 0);
+            }
+            else
+            {
+                entry->mlocked = TRUE;
+            }
+
+            if (!entry->mlocked)
+            {
+                fluid_munlock(entry->sample_data, entry->sample_count * 2);
+                FLUID_LOG(FLUID_WARN, "Failed to pin the sample data to RAM; swapping is possible.");
+            }
+        }
+    }
+
+    entry->num_references++;
+    *sample_data = entry->sample_data;
+    *sample_data24 = entry->sample_data24;
+    ret = FLUID_OK;
+
+unlock_exit:
+    fluid_mutex_unlock(samplecache_mutex);
+    return ret;
+}
+
+int fluid_samplecache_unload(const short *sample_data)
+{
+    fluid_list_t *entry_list;
+    fluid_samplecache_entry_t *entry;
+    int ret;
+
+    fluid_mutex_lock(samplecache_mutex);
+
+    entry_list = samplecache_list;
+    while (entry_list)
+    {
+        entry = (fluid_samplecache_entry_t *)fluid_list_get(entry_list);
+
+        if (sample_data == entry->sample_data)
+        {
+            entry->num_references--;
+
+            if (entry->num_references == 0)
+            {
+                if (entry->mlocked)
+                {
+                    fluid_munlock(entry->sample_data, entry->sample_count * 2);
+                    if (entry->sample_data24 != NULL)
+                    {
+                        fluid_munlock(entry->sample_data24, entry->sample_count);
+                    }
+                }
+
+                fluid_list_remove(samplecache_list, entry);
```

Comment:
This should be `samplecache_list = fluid_list_remove(samplecache_list, entry);`. Otherwise `samplecache_list` is not set to NULL when the last element is removed resulting in a NULL deref in line 248.

Reply:
Of course, well spotted.